### PR TITLE
OGL: Improve performance of bounding box on NVIDIA drivers

### PR DESCRIPTION
**Pull request by stenzek**
Small change that seems to significantly speed up bbox reads on nvidia drivers.

Needs testing on Intel and AMD to ensure nothing is broken.

If this doesn't cause regressions, I'll tidy up the commit and add some documentation.

---
This change is [Reviewable](https://reviewable.io/reviews/dolphin-emu/dolphin/3831)

### CODE REVIEW

**Comment by degasus** [comment]
Code looks good to me, it's just dumb that this speeds up so much.

<details><summary><i>Previously, stenzek (Stenzek) wrote…</i></summary>
> #### \[DO NOT MERGE\] OGL: Improve performance of bounding box on nvidia drivers
> 
> Small change that seems to significantly speed up bbox reads on nvidia drivers.
> 
> 
> 
> Needs testing on Intel and AMD to ensure nothing is broken.
> 
> 
> 
> If this doesn't cause regressions, I'll tidy up the commit and add some documentation.
> 
> 
> </details>


---

Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3831#-:-KHPGHPqnOwY3jhqQNT_:900801566)*

**Comment by JMC47** [comment]
Depending on the internal resolution, this speeds up bounding box from 90 - 300% on my NVIDIA GTX 760.

Tested in the most demanding part of Paper Mario: The Thousand Year Door that I know of - the Battle colosseum while holding flip mode to activate bounding box.

Master - 1x IR = 35 fps
Pull - 1x IR = 63 fps
Master - 4x IR = 19 fps
Pull - 4x IR = 61 fps

**Comment by stenzek** [comment]
https://www.reddit.com/r/opengl/comments/1tulp0/high_performance_reading_of_atomics_or_wtf_does/ seems to be similar to our issue. Doesn't provide an explanation, however.

**Comment by JMC47** [comment]
AMD is unaffected by this change.  While it is awesome at atomics, it still sucks at OpenGL.

**Comment by JMC47** [comment]
Also this is a net negative in lines, which is always a positive ;)

**Comment by degasus** [comment]
Works as fine on mesa/i965

**Comment by degasus** [comment]
So this fixes a performance regression, uses the better GL call, and cleans up code. As it doesn't touch emulation at all, it's also very safe. Fine for 5.0?

**Comment by Helios747** [comment]
Did some quick testing because I had some free time on a haswell ULV.

Peak FPS: 90-120

When bounding box effect hits

Master: 45 FPS

PR: 75 FPS

I'll test this on my desktop that has a haswell when I get home in a few hours.

**Comment by delroth** [comment]
FYI I'm fine with this in 5.0 if you folks make sure it gets good testing coverage on multiple generations of NV GPUs (not just one or two of recent gens).

**Comment by degasus** [comment]
I've tested fermi.

**Comment by JMC47** [comment]
GTX 760, Radeon r9 285 both tested.  GTX 760, as said, is 300% faster when strained by bounding box.  Intel HD was tested by helios.  degasus tested stuff.  This one is really important so we don't have a huge perf regression on Bounding Box with NVIDIA/Intel cards.  I removed it from current since degasus said it wasn't fixable last time and our test build failed to make it faster.

**Comment by mbc07** [comment]
Did a quick test on Fermi-based GT 440 and GT 635M and both were much faster in Paper Mario TTYD boat sequence and in the cutscene after the first battle (where lots of "minions" pops up). I didn't get exact numbers but in master those scenes usually drops to 20-30 FPS at 2x IR here, with this PR they remained at constant 60 FPS...

**Comment by Helios747** [comment]
The following tests were done on a 4790k (HD 4600) at 4.5 ghz and a 980ti on TTYD at 1x IR, and 3x IR + 4x SSAA on Windows 10

(too lazy to pull out broadwell laptop for more extensive tests besides before/after 1x)

**Master**
1x IR

| Card        | max FPS           | Effect FPS|
| ------------- |-------------:| -----:|
| Maxwell (980 ti) | 270 | 10 |
| Haswell (4600)      | 200|  172|
| Broadwell ULV (5500)|110|75|



3x IR + 4x SSAA 

| Card        | max FPS           | Effect FPS|
| ------------- |-------------:| -----:|
| Maxwell (980 ti) | 270 | 2|
| Haswell (4600)      | 45|  32|

**This PR**
1x IR

| Card        | max FPS           | Effect FPS|
| ------------- |-------------:| -----:|
| Maxwell (980 ti) | 270 | 221|
| Haswell (4600)      | 200|  172|
| Broadwell ULV (5500)|110|45|

3x IR + 4x SSAA

| Card        | max FPS           | Effect FPS|
| ------------- |-------------:| -----:|
| Maxwell (980 ti) | 270 | 2|
| Haswell (4600)      | 45|  32|


So, basically this PR is amazing and I don't know how. nvidia saw HUGE gains and Broadwell saw significant gains too, I don't know why Haswell didn't.


So to recap so far, we've tested

- Maxwell
- Kepler
- Fermi
- R9 285 (whatever the heck that is, Tonga?)
- Haswell
- Broadway
- Intel mesa

merge it merge it merge it merge it merge it merge it

**Comment by Helios747** [comment]
@lioncash mentioned that this PR brings his R9 380 down to 11 FPS. @JMC47 this is the same architecture as your card. Can you confirm this?

**Comment by darksylinc** [comment]
I saw this on twitter.. I only want to explain the Delphin devs what is going on:

GL_DYNAMIC_* is like a friggin' suggestion to most GL drivers. Very often they will just stall instead of keeping a triple buffer scheme like you would get in D3D9 (DX11 just doesn't even allow this: you first would need to copy to a StagingBuffer then wait for the StagingBuffer to finish and finally map the StagingBuffer).
 
Another problem with the old code is that calling glMap* needs to sync all the driver's internal threads. That's bad.
That's why GL_PERSISTENT_BIT appeared even though we already had GL_MAP_UNSYNCHRONIZED_BIT (though to be fair GL_MAP_UNSYNCHRONIZED_BIT wasn't legal to be used for reading as per the spec; while persistent is). Persistent mapping didn't remove all sync, but also removed all glMap calls (and thus no driver thread "mini"-stalls)
 
It appears you hit a fast path with the glGetBufferSubData call on the NV driver (also see http://www.stevestreeting.com/2007/03/16/glmapbuffer-how-i-mock-thee/ & http://www.stevestreeting.com/2007/03/17/glmapbuffer-vs-glbuffersubdata-the-return/ they're old but they hold).
 
**Last**: One more thing. I see you're using an SSBO. If you write to the SSBO from a shader, you may need a barrier.
It's very likely you need to call glMemoryBarrier( GL_SHADER_STORAGE_BARRIER_BIT ) before getting the data back (see https://www.opengl.org/wiki/GLAPI/glMemoryBarrier) for the sake of completeness (it's possibly you're getting the wrong value like 0.00001% of the times, or never if lady luck was on your side).
Don't make the barrier call if you didn't modify the data from a shader. On NV it could cripple you again (and the others archs may take a hit as well). May as well add a hack option "ignore barriers on bboxes" if you can't track reliably.
 
On AMD the best performant solution would probably be to use persistent mapping and some fencing; though since emulators aren't in control of what is going to happen next, it may get harder to get an efficient solution quickly; I don't know how often that routine gets called or how it's needed.
 
**Possible Optimization:**
You may want to get rid of one glGenBuffers per BoundingBox; and use one for all BoundingBoxes instead, assigning a region for each BoundingBox (properly aligned as per GL_SHADER_STORAGE_BUFFER_OFFSET_ALIGNMENT).

Though that will not play well with glGetBufferSubData (you'll need persistent mapping and fencing)
It depends on whether you think you can outperform the driver (or at least... provide a consistent implementation that will run more or less the same on all HW. Fences are expensive so don't fence once per BoundingBox::Get)
 
Have fun!

**Comment by AllianceDarkSylinc** [comment]
Forgot to mention: If you go the low level route, don't map directly the SSBO to prevent NV of moving from DEVICE to HOST memory.
Use instead a DX11-like staging approach: Create a buffer you persistently map; and another buffer which is the actual SSBO.

You read/write from/to the staging buffer, and transfer to/from the actual SSBO are handled and enqueue via glCopyBufferData.

On Intel you should avoid the StagingBuffer since this extra copying is pointless (there's no difference between host and device memory). Though I doubt if it's a few MB/s it could pose a performance problem.

Like I said, have fun!

**Comment by darksylinc** [comment]
Forgot to mention: If you go the low level route, don't map directly the SSBO to prevent NV of moving from DEVICE to HOST memory.
Use instead a DX11-like staging approach: Create a buffer you persistently map; and another buffer which is the actual SSBO.

You read/write from/to the staging buffer, and transfer to/from the actual SSBO are handled and enqueued via glCopyBufferData.

On Intel you should avoid the StagingBuffer since this extra copying is pointless (there's no difference between host and device memory). Though I doubt if it's a few MB/s it could pose a performance problem.

Like I said, have fun!

**Comment by JMC47** [comment]
This is actually slower on AMD cards by a good bit in areas limited by bounding box alone.  My original area was hitting a bottleneck somewhere else first, so I didn't notice.

Personally, if I absolutely had to make a choice, I'd say sacrifice the AMD performance for NVIDIA performance (D3D11 is already faster for... everything.) if we absolutely have to.  For now, a driver details switch would probably work so we can get performance back for Non-Win10 NVIDIA users.  Having a huge performance loss for them that makes the games unplayable in the name of accuracy is great an all, but, when there's a solution right here...

**Comment by degasus** [comment]
@darksylinc Yeah, that's what we thought about this issue: Buffer moving. So all further atomic operations on this SSBO will go to the client memory. But I don't get what you mean with "to sync all the driver's internal threads". AFAIK all readback methods have this issue. If you're talking about uploading, yeah, we were hit by this issue hard: https://github.com/dolphin-emu/dolphin/blob/master/Source/Core/VideoBackends/OGL/StreamBuffer.cpp#L204
About the barrier, I thought it's only required to sync between shader interactions. So I thought glGetBufferSubData does handle this well. Am I wrong here?
Do you think the way of readback (persistent staging buffer vs glGetBufferSubData) matters for 16 bytes? There is exactly one bbox, so the amount of memory won't increase.

@JMC47 For a general good performance, we need to "cache" this bbox reads. At least, we need to read back all 4 for of them at once. And cache them till the next draw call.

File: Source/Core/VideoCommon/DriverDetails.cpp [hunk 1]
```diff
@@ -61,6 +61,7 @@ namespace DriverDetails
 		{OS_WINDOWS,VENDOR_NVIDIA,   DRIVER_NVIDIA,   Family::UNKNOWN, BUG_BROKENUNSYNCMAPPING, -1.0, -1.0, true},
 		{OS_LINUX,  VENDOR_NVIDIA,   DRIVER_NVIDIA,   Family::UNKNOWN, BUG_BROKENUNSYNCMAPPING, -1.0, -1.0, true},
 		{OS_WINDOWS,VENDOR_INTEL,    DRIVER_INTEL,    Family::UNKNOWN, BUG_INTELBROKENBUFFERSTORAGE, 101810.3907, 101810.3960, true},
+		{OS_ALL,    VENDOR_NVIDIA,   DRIVER_NVIDIA,   Family::UNKNOWN, BUG_SLOWSSBOMAP, -1.0, -1.0, true},
```
Comment (degasus) [inline Source/Core/VideoCommon/DriverDetails.cpp]: also on windows/intel

File: Source/Core/VideoBackends/OGL/BoundingBox.cpp [hunk 2]
```diff
@@ -42,12 +43,25 @@ int BoundingBox::Get(int index)
...
+		{
+			memcpy(&data, ptr, sizeof(int));
+			glUnmapBuffer(GL_SHADER_STORAGE_BUFFER);
+		}
+	}
+	else
 	{
-		data = *(int*)ptr;
-		glUnmapBuffer(GL_SHADER_STORAGE_BUFFER);
+		// Using glMapBufferRange to read back the contents of the SSBO is extremely slow
```
Comment (degasus) [inline Source/Core/VideoBackends/OGL/BoundingBox.cpp]: Please also tell why: The driver moves the buffer to the cpu memory.

**Comment by stenzek** [comment]
I agree, caching the rectangle is a better way of doing this. D3D12 is implemented this way, but to avoid introducing something that could potentially break, I'm trying to keep this simple.

DriverDetails case added. We now use the glGetBufferSubData path on everything except AMD, which falls back to glMapBufferRange. OS is specified as ANY, as I'm assuming that it's also faster on Linux Catalyst. Can any AMD users please confirm that this no longer causes performance regressions?

**Comment by degasus** [comment]
Reviewed 3 of 3 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3831)*

**Comment by Helios747** [comment]
I'll test this again in a little bit. @JMC47 can you make sure you test the heck out of this on AMD?

**Comment by Parlane** [comment]
@lioncash and @JMC47 please retest this for AMD. If you are no longer affected by this change, this can be merged.

**Comment by JMC47** [comment]
I put in my AMD yet again.  It's not longer affected as far as i can tell.  Maintaining 44 FPS in the opening scene, same as master.

File: Source/Core/VideoCommon/DriverDetails.h [hunk 3]
```diff
@@ -195,6 +195,18 @@ namespace DriverDetails
 		// Mesa introduced geometry shader support for radeon and sandy bridge devices and failed to test it with us.
 		// Causes misrenderings on a large amount of things that draw lines.
 		BUG_BROKENGEOMETRYSHADERS,
+
+		// Bug: glMapBuffer on the SSBO used for bounding box causes significant slowdown
```
Comment (Tilka) [inline Source/Core/VideoCommon/DriverDetails.h]: Please summarize what this flag is used for on AMD, not why the workaround is slow on Nvidia.

**Comment by degasus** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[Source/Core/VideoCommon/DriverDetails.h, line 199 \[r3\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3831#-KHUklejbtW_n67Zo_3M-r3-199:-KI3LbpepegsvDsGbzuS:279000656) ([raw file](https://github.com/dolphin-emu/dolphin/blob/493781540d8f346694d2a53969e3cf3242c6c5ce/Source/Core/VideoCommon/DriverDetails.h#L199)):*
<details><summary><i>Previously, Tilka (flacs) wrote…</i></summary>
> Please summarize what this flag is used for on AMD, not why the workaround is slow on Nvidia.
</details>
Maybe "Likely because of a non-cached access of the four bbox values after each other"?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3831)*

**Comment by nbohr1more** [comment]
Rebase please :)

**Comment by JosJuice** [comment]
@stenzek Please rebase.

**Comment by degasus** [comment]
Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/3831)*
